posts/2021/08/github-actions-benchmarks.md

```diff
@@ -0,0 +1,336 @@
+without reducing the accuracy of the measurements?
+
+# Are CI services reliable enough for benchmarking?
+
+We saw above that CI services are not designed for this kind of tasks, but with some workarounds
+they might be good enough. However, how can be sure? As data-driven scientists, we say let's run
+an experiment!
+
+## The setup
+
+Our experiment is data-drive, so we need to generate some data first. This is our strategy:
+
+* We will benchmark and compare two commits that are exactly the same in terms of tested code.
+* Under ideal conditions, we should see that the performance ratio between the two commits is 1.0.
+In other words, performance should be the same. Of course, these are not ideal conditions, so
+some kind of error is expected. We want to see if it stays reliably under an acceptable threshold.
+* We will implement a GitHub Actions workflow that will run every six hours (four times a day) and
+collect results for a week or more. This will help us account for two things:
+    * Accumulate sufficient data points to answer our question.
+    * Account for time factors like different days of the week (e.g. weekday vs weekend),
+      or the time of the day (3am vs 3pm).
+* The workflow will upload the benchmark results as artifacts we can download and process locally
+  with a Jupyter Notebook.
+
+> Check the [GitHub Actions workflow in this fork](https://github.com/jaimergp/scikit-image/blob/main/.github/workflows/benchmarks-cron.yml)!
+
+## The results
+
+To measure the stability of the different benchmark runs we will be looking at the performance
+ratio of the measurements between the two commits. Since they are identical in terms of tested code,
+ideally they should be all `1.0`. We know this will not happen, but maybe the errors are not that big
+and stay like that regardless the time of the day or the day of the week.
+
+After some days collecting data points, these are the results:
+
+![Reliability of benchmarks in GitHub Actions](/images/2021/08/github-actions-benchmark.png)
+
+* Average time taken: 1h55min
+* Minimum and maximum ratios observed: 0.51, 1.36
+* Mean and standard deviation of ratios: 1.00, 0.05
+* Proportion of false positives: 4/108 = 3.7%
+
+In the X axis you can see the different runs, sorted by date and time. Days of the week are grouped
+with colored patches for easier visual analysis. In the Y axis, we plot the performance ratio. Each
+of those vertical clouds include 75 points, one per benchmark test. Ideally, they should all fall at
+`y=1`. Of course, not all of them are there, but a surprising amount of them do!
+
+But! We do see some bigger deviations at certain times! How is that possible? Well, that's our error!
+These outliers are the ones that would concern us because they can be interpreted as **false positives**.
+In other words: `asv` would report a performance regression, when in fact there's none. However,
+in the observed measurements, the outliers were always within `y ∈ (0.66, 1.5)`.
```

Comment:
This line states 0.66, but the "minimum" that was stated a paragraph or two before was 0.51. seems like a contradiction?

Reply:
You are right, let me look into this!